fs-index/src/index.rs

```diff
@@ -1,1120 +1,501 @@
+/// use dev_hash::Crc32;
+///
+/// // Define the root path
+/// let root_path = Path::new("animals");
+///
+/// // Build the index
+/// let index: ResourceIndex<Crc32> = ResourceIndex::build(root_path).expect("Failed to build index");
+/// // Store the index
+/// index.store().expect("Failed to store index");
+///
+/// // Load the stored index
+/// let mut loaded_index: ResourceIndex<Crc32> = load_or_build_index(root_path, false).expect("Failed to load index");
+///
+/// // Update the index
+/// loaded_index.update_all().expect("Failed to update index");
+///
+/// // Get a resource by path
+/// let _resource = loaded_index
+///     .get_resource_by_path("cat.txt")
+///     .expect("Resource not found");
+///
+/// // Track the removal of a file
+/// loaded_index
+///     .track_removal(Path::new("cat.txt"))
+///     .expect("Failed to track removal");
+///
+/// // Track the addition of a new file
+/// loaded_index
+///     .track_addition(Path::new("dog.txt"))
+///     .expect("Failed to track addition");
+///
+/// // Track the modification of a file
+/// loaded_index
+///     .track_modification(Path::new("dog.txt"))
+///     .expect("Failed to track modification");
+/// ```
+#[derive(Clone, Debug)]
+pub struct ResourceIndex<Id>
 where
-    Id: ResourceId,
+    Id: Eq + Hash,
 {
-    entries
-        .into_iter()
-        .filter_map(|(path_buf, entry)| {
-            let metadata = entry.metadata().ok()?;
-
-            let path = path_buf.as_canonical_path();
-            let result = scan_entry(path, metadata);
-            match result {
-                Err(msg) => {
-                    log::error!(
-                        "Couldn't retrieve metadata for {}:\n{}",
-                        path.display(),
-                        msg
-                    );
-                    None
-                }
-                Ok(entry) => Some((path_buf, entry)),
-            }
-        })
-        .collect()
+    /// The root path of the index (canonicalized)
+    pub(crate) root: PathBuf,
+    /// A map from resource IDs to resources
+    ///
+    /// Multiple resources can have the same ID (e.g., due to hash collisions
+    /// or files with the same content)
+    pub(crate) id_to_resources: HashMap<Id, Vec<IndexedResource<Id>>>,
+    /// A map from resource paths to resources
+    pub(crate) path_to_resource: HashMap<PathBuf, IndexedResource<Id>>,
 }
 
-fn is_hidden(entry: &DirEntry) -> bool {
-    entry
-        .file_name()
-        .to_str()
-        .map(|s| s.starts_with('.'))
-        .unwrap_or(false)
+/// Represents the result of an update operation on the ResourceIndex
+#[derive(PartialEq, Debug)]
+pub struct IndexUpdate<Id: ResourceId> {
+    /// Resources that were added during the update
+    added: Vec<IndexedResource<Id>>,
+    /// Resources that were modified during the update
+    modified: Vec<IndexedResource<Id>>,
+    /// Resources that were removed during the update
+    removed: Vec<IndexedResource<Id>>,
```

Comment:
Is `last_modified` timestamp correct in items of `removed` vector?

It's time of last modification before deletion, not time of removal, right?

Reply:
It's the time of the last modification.

The `IndexedResource` fields are only updated when we call the `update_all()` or `track_*()` methods.

Reply:
I don't think that we should provide users with this timestamp for removed items. It might be confusing. But `modified` entry is confusing to me in either case, because resources are immutable. They can be only added or removed, their locations are secondary. We might support `moved` event though.

In future versions of the API, I would expect events like `new version` for cases when one resource is derived from another.

Reply:
Changed the return value of `ResourceIndex::update_all()` to include only two maps: added and removed.

Reply:
Let's also use `HashSet` instead of `Vec` to ensure that duplicates are not included. Order of elements should not matter.

Reply:
And we don't need to provide users with `last_modified` timestamps of removed resources, because they are outdated and do not reflect time of removal. So it's confusing and may provoke misuse.

Also, we should not provide users with paths of removed resources. There is no use for them. If we keep paths of removed resources in our API, users could think that they should update elements by these paths. But this is job for `fs-index`, users only need to track resource ids, read content by path when it is updated (but not when removed) and use timestamps for sortings.

So, we should only emit ids of removed resources.
```
pub struct IndexUpdate<Id: ResourceId> {
    /// Resources that were added during the update
    added: HashSet<IndexedResource<Id>>,
    
    /// Resources that were removed during the update
    removed: HashSet<Id>,
}
```

Reply:
Lastly, we should maintain invariant that `added` set never contains elements with same ids and different paths/timestamps. This can be expressed in the code:
```
pub struct IndexEntry {
    /// The path of the resource, relative to the root path
    path: PathBuf,

    /// The last modified time of the resource (from the file system metadata)
    last_modified: SystemTime,
}

pub struct IndexUpdate<Id: ResourceId> {
    /// Resources that were added during the update
    added: HashMap<Id, IndexEntry>,

    /// Resources that were removed during the update
    removed: HashSet<Id>
}
```

Reply:
> users only need to track resource ids, read content by path when it is updated

So, I assume we need `IndexUpdate` to be:
```rust
/// Represents the result of an update operation on the ResourceIndex
#[derive(PartialEq, Debug)]
pub struct IndexUpdate<Id: ResourceId> {
    /// Resources that were added during the update
    added: HashMap<Id, IndexedResource<Id>>,
    /// Resources that were removed during the update
    removed: HashSet<Id>,
}
```
Where `added` is a map between `Id` and `IndexedResource` (Id, path, last modified time), not `IndexEntry` (Id, last modified time).
right?